bpf: create and open maps using `cilium/ebpf`

ti-mo opened the pull request:
This PR puts `cilium/ebpf.Map` at the core of Cilium's `bpf.Map` type, see #22513 for more context. Best reviewed per commit given its size. If you've been tagged for review and don't feel comfortable, please let me know so we can reassign. Don't hesitate to ping me offline for discussion or more context.

Part of: #22513.

```release-note
Perform map creation and opening using cilium/ebpf API
```

ti-mo committed:
bpf: remove MapFdFromID, BPF_MAP/PROG/OBJ commands for bpf()

MapFdFromID was unused in the codebase, remove it. Also remove
unused command args to bpf(). The remainder acts as a list of behaviour
to port to use cilium/ebpf.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/bpf: replace BPF_OBJ_GET syscall with cilium/ebpf

OpenMap() now calls ebpf.LoadPinnedMap and must be given an asbolute path.
bpf.ObjGet() was removed as it is now unused.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: Map: remove Parallel mode and NonPersistent in favor of Recreate()

A map is considered in 'parallel' mode when it has been unpinned, recreated and
re-pinned, but the old map is still referred to by living endpoint programs until
those eventually get phased out by replacing their progs.

Rename OpenParallel() to Recreate() as it's a confusing term from a data/bpf
standpoint, as there are no double-writes going on.

The Map.NonPersistent flag is a deferred way of saying "always remove the map's
pin when creating". Remove this bit of state, instead explicitly calling
Recreate() for opening the Map.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/bpf: remove Map.fd in favor of ebpf.Map

- Remove the bpf.Map.fd field and all accesses to it
- Rename bpf.Map.GetFd() to FD() and perform all map FD accesses through it

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
daemon: decorate all map creation errors

Errors returned from OpenOrCreateMap don't necessarily contain any identifying
context about the map being created. Wrap all errors encountered during map
creation with the map/subsystem name.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove bool return from OpenOrCreate()

This commit removes the bool return value from OpenOrCreate() that indicates
whether or not a Map was recreated in the process.

It was only checked in tests and generally didn't provide much value or
actionable feedback, only inflating API surface.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/bpf: replace BPF_MAP_CREATE/OBJ_PIN syscalls with cilium/ebpf

This is the first in a series of patches that aims to remove all uses of
unix.Syscall in the cilium/cilium codebase in favour of the cilium/ebpf APIs.

CreateMap() was unexported and renamed to createMap(). It now calls directly into
ebpf.NewMapWithOptions and takes care of passing MapOptions when specified (for
passing a PinPath).

OpenOrCreateMap() now takes an ebpf.MapSpec and a pinDir and had most of its logic
removed as compatibility checks are already taken care of by ebpf-go.

NewMap() no longer takes innerID. A new function NewMapWithInnerSpec() was introduced
to provide an ebpf.MapSpec to use for the inner map definition during map creation.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/bpf: remove unused innerID argument from NewMap()

The innerID argument was only used by 2 maps that needed inner maps. Those were
migrated to NewMapWithInnerSpec() in a prior commit, when the argument was
already blanked out. This commit removes it.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
ctmap: make PerClusterCTMap.getClusterMap return error for nonexistent map

By changing the underlying map abstraction to cilium/ebpf, opening a map from a pin
no longer returns an error chain containing a os.PathError, only a unix.ENOENT.
Return errors from getClusterMap() instead of a vague 'nil, nil'.

Check for ENOENT in getAllClusterMaps() instead of checking the first return value.
Stop checking the first return value in multiple places.

Also removed redundant mapType argument from PerClusterCTMap.newInnerMap().

Update tests to perform error checking first before accessing any other return value.

Signed-off-by: Timo Beckers <timo@isovalent.com>

YutaroHayakawa reviewed (approved):
Sorry for the late response 🙇 

Thanks for taking care of the per-cluster maps. The changes looks good. Wrt the pinned maps, I totally agree that it introduces the complexity, but the time takes to iterate over the outer maps is significantly longer when we don't pin the inner map. For example, measurement in my local machine shows 5s for iterating single per-cluster map without pin and 600us with pin. Thus, with pin, 20s per GC.

I'm worrying about the scenario that users have clusters with very high CT fill-up rate and want to reduce the GC rate, this latency becomes the problem. However, I don't know how short do people setup the CT GC practically. The default GC interval is 5m, so in default scenario (ignoring the auto interval adjustment), indeed, it doesn't become a big problem.

dylandreimerink reviewed (approved):
LGTM. Thanks for the clean history and small commits, makes it easy to review :rocket:

lmb commented on pkg/bpf/bpf_linux.go:
```diff
@@ -460,82 +354,60 @@ func objCheck(fd int, path string, mapType MapType, keySize, valueSize, maxEntri
+			}).Info("Unpinning map with incompatible properties")
+
+		// Found pinned map incompatible with spec. Remove it and recreate the map.
+		if err := os.RemoveAll(path.Join(pinDir, spec.Name)); err != nil {
```
Nit: why not `m.Unpin()`?

ti-mo replied:
Of course, thanks!

lmb commented on pkg/maps/cidrmap/cidrmap.go:
```diff
@@ -140,50 +142,49 @@ func (cm *CIDRMap) Close() error {
+		ValueSize:  uint32(LPM_MAP_VALUE_SIZE),
+		MaxEntries: maxelem,
+		Flags:      bpf.BPF_F_NO_PREALLOC,
+		Pinning:    1,
```
Nit: use the named constant?

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -71,7 +71,6 @@ type MapInfo struct {
 	MaxEntries    uint32
 	Flags         uint32
 	InnerSpec     *ebpf.MapSpec
-	OwnerProgType ProgType
```
I guess this was unused?

ti-mo replied:
Yep, wasn't used.

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -402,38 +397,32 @@ func GetMapInfo(pid int, fd int) (*MapInfo, error) {
+			MapType:       MapType(em.Type()),
+			KeySize:       em.KeySize(),
+			ValueSize:     em.ValueSize(),
+			ReadValueSize: em.ValueSize(),
```
Not sure whether this is actually used in the code base, but this value is going to be wrong for per CPU maps, no?

ti-mo replied:
Hmm, good catch. We used to have [`NewPerCPUHashMap`](https://github.com/cilium/cilium/commit/60eef6f60ce74b659ac493a593f24f7fe8b3a15d), but this has since been removed. I replicated `GetMapInfo()` behaviour here, which also populates ReadValueSize with the `value_size:%d` value found in a map's fdinfo.

As far as I can see, `METRICS_MAP` is still in use and is a per-cpu hash map. I assume fdinfo echoes the full value with across all CPUs, otherwise dumping metrics (e.g. using `cilium metrics list` from within the container) would currently be broken if the metrics map was opened from a pin. I'll need to double-check on Monday.

ti-mo replied:
Nope, fdinfo of percpu hashes shows the value size given during creation, not the total over all CPUs. Turns out `pkg/maps/metricsmap.metricsMap` used by `cilium bpf metrics list` is already implemented on top of `pkg/ebpf` using ebpf-go.

Given the code here replaces `GetMapInfo` behaviour 1:1, this is not going to break anything. When we implement `Map.DumpWithCallback` and friends on top of `Map.Iterate()`, we'll add support for per-cpu maps again. I've left a warning in the method godoc to remind us.

lmb commented on pkg/maps/tunnel/tunnel.go:
```diff
@@ -60,7 +60,7 @@ func NewTunnelMap(mapName string) *Map {
 		0,
 		bpf.ConvertKeyValue,
-	).WithCache().WithPressureMetric().WithNonPersistent().
+	).WithCache().WithPressureMetric().
```
So where is the matching `Recreate` call?

ti-mo replied:
See `tunnel.TunnelMap().Recreate()` and `vtep.VtepMap().Recreate()` in `daemon/cmd/datapath.go` for both 'NonPersistent' maps.

lmb commented on pkg/bpf/bpf_linux.go:
```diff
@@ -42,6 +42,33 @@ func createMap(spec *ebpf.MapSpec, opts *ebpf.MapOptions) (*ebpf.Map, error) {
+		return nil, fmt.Errorf(
+			"incompatible map at path %s: old: (Type:%s KeySize:%d ValueSize:%d MaxEntries:%d, Flags:%d), "+
+				"new: (Type:%s KeySize:%d ValueSize:%d MaxEntries:%d, Flags:%d): %w", path,
+			info.Type, info.KeySize, info.ValueSize, info.MaxEntries, info.Flags,
```
Doesn't `m.Type()` etc. should give you what you need?

ti-mo replied:
Thanks, forgot those existed!

lmb commented on pkg/bpf/bpf_linux.go:
```diff
@@ -42,6 +42,33 @@ func createMap(spec *ebpf.MapSpec, opts *ebpf.MapOptions) (*ebpf.Map, error) {
+			return nil, fmt.Errorf("get info of incompatible map: %w", err)
+		}
+
+		return nil, fmt.Errorf(
```
FWIW we could extend the library to return an more useful error. Or create a dedicated error type which contains the info needed.

ti-mo replied:
Yes, that might be better. Would like to merge this PR in the short term, though, so we can keep this here for now. Tracking this in https://github.com/cilium/ebpf/issues/1034.

lmb commented on pkg/maps/policymap/policymap.go:
```diff
@@ -453,7 +455,15 @@ func newMap(path string) *PolicyMap {
+
 	err := m.OpenOrCreate()
+	if errors.Is(err, ebpf.ErrMapIncompatible) {
+		err = m.Recreate()
```
Shouldn't this be part of the previous commit?

ti-mo replied:
This commit has been removed, but the goal was to split off the change to the building block from the change to the consumer, as it might be easier to review that way. In hindsight, probably better to put them together to avoid mistakes.

lmb reviewed (commented):
Thanks a lot for breaking this down into individual commits, it really helps with review!

nebril reviewed (approved):
ci-structure bit LGTM

tommyp1ckles commented on pkg/bpf/bpf_linux.go:
```diff
@@ -460,82 +354,55 @@ func objCheck(fd int, path string, mapType MapType, keySize, valueSize, maxEntri
+		}
+
+		if err := os.MkdirAll(pinDir, 0755); err != nil {
+			return nil, fmt.Errorf("creating map base pinning directory: %w", err)
```
nit: can we add the pinDir to the error message to aid in debugging.

ti-mo replied:
That's on purpose, errors returned from os.Mkdir and friends already include the path: https://go.dev/play/p/RZbMjf2Q55z.

```go
func main() {
	fmt.Println(os.MkdirAll("/proc/foo", 0777))
}

mkdir /proc/foo: operation not permitted
```

tommyp1ckles commented on pkg/bpf/bpf_linux.go:
```diff
@@ -460,82 +354,55 @@ func objCheck(fd int, path string, mapType MapType, keySize, valueSize, maxEntri
+
+		log.WithField(logfields.Path, path.Join(pinDir, spec.Name)).
+			WithFields(logrus.Fields{
+				"old": fmt.Sprintf("Type:%s KeySize:%d ValueSize:%d MaxEntries:%d, Flags:%d",
```
question - why not just use the logrus kv API?

ti-mo replied:
This may be personal preference, but I find the kv approach harder to visually diff: https://go.dev/play/p/xlpzqnxCIig.

```
time="2009-11-10T23:00:00Z" level=info msg="Unpinning map with incompatible properties" new="Type:Array KeySize:4 ValueSize:4 MaxEntries:1024 Flags:0" old="Type:Hash KeySize:4 ValueSize:4 MaxEntries:1024 Flags:0"
```
vs
```
time="2009-11-10T23:00:00Z" level=info msg="Unpinning map with incompatible properties" new_flags=0 new_keysize=4 new_maxentries=4 new_type=Array new_valuesize=4 old_flags=0 old_keysize=4 old_maxentries=4 old_type=Hash old_valuesize=4
```

- Lots of repetition (old_ and new_ prefixes)
- Longer
- Fields get sorted alphabetically
- The convention is to lowercase keys, which sort of breaks the correspondence to the keys in the MapSpec

tommyp1ckles replied:
I see what you mean, the main advantage of the standard way is that it might hook better into existing log aggregation services.

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -155,35 +159,36 @@ func NewMap(name string, mapType MapType, mapKey MapKey, keySize int,
+func NewMapWithInnerSpec(name string, mapType MapType, mapKey MapKey, mapValue MapValue, maxEntries int, flags uint32,
+	innerSpec *ebpf.MapSpec, dumpParser DumpParser) *Map {
+
+	keySize := reflect.TypeOf(mapKey).Elem().Size()
```
note: Calls to `Elem()` can panic, such as if mapKey is not a pointer type.

ti-mo replied:
Ah, good observation, didn't realize. It's already been in use in NewMap for years, so we can live with this for now. The goal is to make bpf.Map generic in the short term, e.g. `Map[K MapKey, V MapValue]` so sizing can be done at compile time.

ti-mo committed:
Makefile: add ECHO_CHECK statements to integration and privileged tests

When running `V=0 make ... --quiet`, the `go test ./...` executions will never
show up, and all output is redirected to /dev/null. This makes it hard to see
what's going on.

On Travis, this can cause timeouts due to the job emitting no output.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
.travis: increase verbosity of 'make integration-tests'

We're having issues with Travis jobs timing out due to no output being
generated for over 10 minutes.

Integration tests run with V=0 and --quiet barely generate any output,
since all 'go test' output is hidden, and Makefile steps are also hidden.

Signed-off-by: Timo Beckers <timo@isovalent.com>

squeed reviewed (approved):
approved the tophat-relevant files.
